libavutil/hwcontext_mediafoundation.c

```diff
@@ -0,0 +1,445 @@
+
+static void mf_uninit_d3d(AVHWDeviceContext *ctx)
+{
+    AVMFDeviceContext *hwctx = ctx->hwctx;
+    MFDeviceContext *priv = ctx->internal->priv;
+
+    if (hwctx->d3d11_manager)
+        IMFDXGIDeviceManager_Release(hwctx->d3d11_manager);
+    hwctx->d3d11_manager = NULL;
+
+    if (priv->d3d11_dll)
+        FreeLibrary(priv->d3d11_dll);
+    priv->d3d11_dll = NULL;
+
+    if (hwctx->d3d9_manager)
+        IDirect3DDeviceManager9_Release(hwctx->d3d9_manager);
+    hwctx->d3d9_manager = NULL;
+
+    if (priv->d3d9)
+        IDirect3D9_Release(priv->d3d9);
+    priv->d3d9 = NULL;
+
+    if (priv->d3d9_dll)
+        FreeLibrary(priv->d3d9_dll);
+    priv->d3d9_dll = NULL;
+
+    if (priv->dxva2_dll)
+        FreeLibrary(priv->dxva2_dll);
+    priv->dxva2_dll = NULL;
+}
+
+static int mf_create_d3d11_device(AVHWDeviceContext *ctx, int loglevel)
+{
+    AVMFDeviceContext *hwctx = ctx->hwctx;
+    MFDeviceContext *priv = ctx->internal->priv;
+    HRESULT hr;
+    UINT token;
+    ID3D11Device *d3d11_device = NULL;
+    ID3D10Multithread *multithread;
+    HRESULT (WINAPI *pD3D11CreateDevice)(
+    _In_opt_        IDXGIAdapter        *pAdapter,
+                    D3D_DRIVER_TYPE     DriverType,
+                    HMODULE             Software,
+                    UINT                Flags,
+    _In_opt_  const D3D_FEATURE_LEVEL   *pFeatureLevels,
+                    UINT                FeatureLevels,
+                    UINT                SDKVersion,
+    _Out_opt_       ID3D11Device        **ppDevice,
+    _Out_opt_       D3D_FEATURE_LEVEL   *pFeatureLevel,
+    _Out_opt_       ID3D11DeviceContext **ppImmediateContext
+    );
+    HRESULT (WINAPI *pMFCreateDXGIDeviceManager)(
+    _Out_ UINT                 *pResetToken,
+    _Out_ IMFDXGIDeviceManager **ppDXVAManager
+    );
+    HANDLE mfplat_dll;
+
+    if (hwctx->init_d3d11_device) {
+        d3d11_device = hwctx->init_d3d11_device;
+        ID3D11Device_AddRef(d3d11_device);
+    } else {
+        priv->d3d11_dll = LoadLibraryW(L"D3D11.dll");
+        if (!priv->d3d11_dll)
+            return AVERROR_EXTERNAL;
+
+        pD3D11CreateDevice = (void *)GetProcAddress(priv->d3d11_dll, "D3D11CreateDevice");
+        if (!pD3D11CreateDevice)
+            return AVERROR_EXTERNAL;
+
+        hr = pD3D11CreateDevice(0,
+                                D3D_DRIVER_TYPE_HARDWARE,
+                                NULL,
+                                D3D11_CREATE_DEVICE_VIDEO_SUPPORT,
+                                NULL,
+                                0,
+                                D3D11_SDK_VERSION,
+                                &d3d11_device,
+                                NULL,
+                                NULL);
+        if (FAILED(hr)) {
+            av_log(ctx, loglevel, "failed to create D3D device\n");
+            goto error;
+        }
+
+        hr = IMFMediaBuffer_QueryInterface(d3d11_device, &IID_ID3D10Multithread, (void **)&multithread);
```

Comment:
If we use version 11, we can change this line and the ones that follow.

Reply:
I will have this in my next commit.  Thanks :)